### PR TITLE
Many fixes

### PR DESCRIPTION
**Pull request by samarion**
Fixes include:
- (partial) support for well-formedness check on free variables from @mantognini 
- refactored the logic in termination pipeline runner
- fixed and generalized handling of pre/post/measure specs during transformation pipeline
  (fixes https://github.com/epfl-lara/stainless/issues/114 but still needs docs) 
- more stable parallelism (still not done though)

### CODE REVIEW

File: core/src/main/scala/stainless/extraction/oo/Definitions.scala [hunk 1]
```diff
@@ -59,44 +59,72 @@ trait Definitions extends extraction.Trees { self: Trees =>
   }
 
   case class TypedClassDef(cd: ClassDef, tps: Seq[Type])(implicit val symbols: Symbols) extends Tree {
-    lazy val id: Identifier = cd.id
+    @inline def id: Identifier = cd.id
 
-    lazy val typeMap: Map[TypeParameter, Type] = (cd.typeArgs zip tps).toMap
+    private[this] var _typeMap: Map[TypeParameter, Type] = _
```
Comment (mantognini) [inline core/src/main/scala/stainless/extraction/oo/Definitions.scala]: I noticed you introduced `Lazy` in inox but you don't use it here. Any reasons for this choice?

Comment (samarion) [reply core/src/main/scala/stainless/extraction/oo/Definitions.scala]: Forgetfulness...

Comment (mantognini) [reply core/src/main/scala/stainless/extraction/oo/Definitions.scala]: That's a valid reason. ;-)

**Comment by samarion** [comment]
Shall we go ahead and merge this and fix the remaining parallelism issues in another PR?

**Comment by mantognini** [comment]
Sounds good to me.

**Pull request merged**